File: src/utils/money.ts
/**
 * Money handling for the optional booking amount.
 *
 * Amounts are stored as an INTEGER number of minor units (paise), never as a
 * float: 0.1 + 0.2 problems have no place in a record of what a guest paid.
 *
 * Formatting is done by hand rather than through Intl so the output is identical
 * on every device and in tests — ICU data differs between Android versions, and
 * grouping is the one thing that would silently change.
 */

/** Change this one constant to use a different currency. */
export const CURRENCY_SYMBOL = '₹';

/** Two decimal places, i.e. 100 paise to the rupee. */
export const MINOR_UNITS_PER_MAJOR = 100;

/** Generous ceiling that still catches a slipped decimal point or typo. */
export const MAX_AMOUNT_MINOR = 100_000_000; // 1,000,000.00

/**
 * Parses user input into minor units.
 * Returns null for blank input (the field is optional).
 * Throws nothing — invalid input yields `undefined`, which callers treat as an error.
 */
export function parseAmountToMinor(input: string): number | null | undefined {
  const cleaned = input
    .replace(CURRENCY_SYMBOL, '')
    .replace(/[\s,]/g, '')
    .trim();
  if (cleaned === '') return null;

  // Optional single decimal point with at most two digits after it.
  if (!/^\d+(\.\d{1,2})?$/.test(cleaned)) return undefined;

  const [whole, fraction = ''] = cleaned.split('.');
  const minor =
    Number(whole) * MINOR_UNITS_PER_MAJOR + Number(fraction.padEnd(2, '0'));
  if (!Number.isSafeInteger(minor) || minor > MAX_AMOUNT_MINOR) return undefined;
  return minor;
}

/** Groups digits in threes, e.g. 1234567 -> "1,234,567". */
function groupThousands(digits: string): string {
  let out = '';
  for (let i = 0; i < digits.length; i += 1) {
    if (i > 0 && (digits.length - i) % 3 === 0) out += ',';
    out += digits[i];
  }
  return out;
}

/** `₹1,500.00`, or an em dash when there is no amount. */
export function formatMinor(minor: number | null | undefined): string {
  if (minor === null || minor === undefined || !Number.isFinite(minor)) return '—';
  const negative = minor < 0;
  const abs = Math.abs(Math.round(minor));
  const major = Math.floor(abs / MINOR_UNITS_PER_MAJOR);
  const fraction = String(abs % MINOR_UNITS_PER_MAJOR).padStart(2, '0');
  return `${negative ? '-' : ''}${CURRENCY_SYMBOL}${groupThousands(String(major))}.${fraction}`;
}

/** Value for an editable text field: "1500.00", or '' when unset. */
export function minorToInput(minor: number | null | undefined): string {
  if (minor === null || minor === undefined || !Number.isFinite(minor)) return '';
  const abs = Math.round(minor);
  const major = Math.floor(abs / MINOR_UNITS_PER_MAJOR);
  return `${major}.${String(abs % MINOR_UNITS_PER_MAJOR).padStart(2, '0')}`;
}

/** Field-level validation message, or undefined when acceptable (incl. blank). */
export function validateAmount(input: string): string | undefined {
  const parsed = parseAmountToMinor(input);
  if (parsed === undefined) {
    return `Enter an amount like 1500 or 1500.50 (up to ${formatMinor(MAX_AMOUNT_MINOR)})`;
  }
  return undefined;
}

File: lib/utils/money.dart
/// Minor-unit (paise) currency helpers — mirrors `src/utils/money.ts`.
///
/// All amounts are stored as integer minor units, never floats, to avoid
/// float rounding issues (the classic 0.1 + 0.2 problem) in a record of what
/// a guest paid.
library;

const currencySymbol = '₹';
const _minorUnitsPerMajor = 100;
/// Generous ceiling (₹1,000,000.00) that still catches a slipped decimal or typo.
const maxAmountMinor = 100000000;

/// Three-state parse result for an amount text field.
///
/// - [valid] true & [minor] null: field left blank — a legitimate "not
///   recorded" amount.
/// - [valid] true & [minor] non-null: a parsed amount.
/// - [valid] false: the text could not be parsed as a valid amount.
class ParsedAmount {
  final int? minor;
  final bool valid;
  const ParsedAmount.value(this.minor) : valid = true;
  const ParsedAmount.invalid()
      : minor = null,
        valid = false;
}

final _amountPattern = RegExp(r'^\d+(\.\d{1,2})?$');

/// Parses free-form text (currency symbol/commas/whitespace tolerated) into
/// minor units.
ParsedAmount parseAmountToMinor(String input) {
  var cleaned = input.trim();
  if (cleaned.isEmpty) return const ParsedAmount.value(null);
  cleaned = cleaned.replaceAll(currencySymbol, '').replaceAll(',', '').replaceAll(RegExp(r'\s+'), '');
  if (cleaned.isEmpty) return const ParsedAmount.value(null);
  if (!_amountPattern.hasMatch(cleaned)) return const ParsedAmount.invalid();
  final parts = cleaned.split('.');
  final whole = parts[0];
  final fraction = parts.length > 1 ? parts[1].padRight(2, '0') : '00';
  final wholeValue = int.tryParse(whole);
  final fractionValue = int.tryParse(fraction);
  if (wholeValue == null || fractionValue == null) return const ParsedAmount.invalid();
  final minor = wholeValue * _minorUnitsPerMajor + fractionValue;
  if (minor > maxAmountMinor) return const ParsedAmount.invalid();
  return ParsedAmount.value(minor);
}

/// Manual thousands-comma grouping — deliberately not using `intl`'s
/// `NumberFormat`, whose ICU data can differ across Android versions and
/// would make formatting non-deterministic across devices.
String _groupThousands(String digits) {
  final buffer = StringBuffer();
  final offset = digits.length % 3;
  for (var i = 0; i < digits.length; i++) {
    if (i != 0 && (i - offset) % 3 == 0) buffer.write(',');
    buffer.write(digits[i]);
  }
  return buffer.toString();
}

/// `'₹1,500.00'`, or `'—'` if [minor] is null.
String formatMinor(int? minor) {
  if (minor == null) return '—';
  final negative = minor < 0;
  final abs = minor.abs();
  final whole = abs ~/ _minorUnitsPerMajor;
  final fraction = (abs % _minorUnitsPerMajor).toString().padLeft(2, '0');
  final grouped = _groupThousands(whole.toString());
  return '${negative ? '-' : ''}$currencySymbol$grouped.$fraction';
}

/// Plain `'1500.00'` (no symbol/grouping) for an editable text field.
String minorToInput(int? minor) {
  if (minor == null) return '';
  final whole = minor ~/ _minorUnitsPerMajor;
  final fraction = (minor % _minorUnitsPerMajor).toString().padLeft(2, '0');
  return '$whole.$fraction';
}

/// Field-level validation message, or null if [input] is valid (including blank).
String? validateAmount(String input) {
  final parsed = parseAmountToMinor(input);
  if (!parsed.valid) {
    return 'Enter a valid amount (up to 2 decimal places).';
  }
  if (parsed.minor != null && parsed.minor! > maxAmountMinor) {
    return 'Amount must be at most ${formatMinor(maxAmountMinor)}.';
  }
  return null;
}
